ignition_ros2_control/src/ignition_ros2_control_plugin.cpp

```diff
@@ -0,0 +1,400 @@
+
+  if (!this->dataPtr->controller_manager_->has_parameter("update_rate")) {
+    RCLCPP_ERROR_STREAM(
+      this->dataPtr->node->get_logger(),
+      "controller manager doesn't have an update_rate parameter");
+    return;
+  }
+
+  auto worldEntity =
+    _ecm.EntityByComponents(ignition::gazebo::components::World());
+
+  auto physicsComp =
+    _ecm.Component<ignition::gazebo::components::Physics>(worldEntity);
+  const auto & physicsParams = physicsComp->Data();
+  const auto newStepSize =
+    std::chrono::duration<double>(physicsParams.MaxStepSize());
+
+  // Get the Gazebo simulation period
+  rclcpp::Duration gazebo_period(
+    std::chrono::duration_cast<std::chrono::nanoseconds>(
+      std::chrono::duration<double>(newStepSize)));
+
+  auto cm_update_rate = this->dataPtr->controller_manager_->get_parameter("update_rate").as_int();
+  this->dataPtr->control_period_ = rclcpp::Duration(
+    std::chrono::duration_cast<std::chrono::nanoseconds>(
+      std::chrono::duration<double>(1.0 / static_cast<double>(cm_update_rate))));
+  // Check the period against the simulation period
+  if (this->dataPtr->control_period_ < gazebo_period) {
+    RCLCPP_ERROR_STREAM(
+      this->dataPtr->node->get_logger(),
+      "Desired controller update period (" << this->dataPtr->control_period_.seconds() <<
+        " s) is faster than the gazebo simulation period (" <<
+        gazebo_period.seconds() << " s).");
+  } else if (this->dataPtr->control_period_ > gazebo_period) {
+    RCLCPP_WARN_STREAM(
+      this->dataPtr->node->get_logger(),
+      " Desired controller update period (" << this->dataPtr->control_period_.seconds() <<
+        " s) is slower than the gazebo simulation period (" <<
+        gazebo_period.seconds() << " s).");
+  }
+
+  this->dataPtr->entity_ = _entity;
+}
+
+//////////////////////////////////////////////////
+void IgnitionROS2ControlPlugin::Update(
+  const ignition::gazebo::UpdateInfo & _info,
+  ignition::gazebo::EntityComponentManager & /*_ecm*/)
+{
+  // Get the simulation time and period
+  rclcpp::Time sim_time_ros(std::chrono::duration_cast<std::chrono::nanoseconds>(
+      _info.simTime).count());
+  rclcpp::Duration sim_period = sim_time_ros - this->dataPtr->last_update_sim_time_ros_;
+
+  if (sim_period >= this->dataPtr->control_period_) {
+    this->dataPtr->last_update_sim_time_ros_ = sim_time_ros;
+    this->dataPtr->controller_manager_->read();
+    this->dataPtr->controller_manager_->update();
+    this->dataPtr->controller_manager_->write();
+  }
+}
+}  // namespace ignition_ros2_control
+
+IGNITION_ADD_PLUGIN(
+  ignition_ros2_control::IgnitionROS2ControlPlugin,
+  ignition::gazebo::System,
+  ignition_ros2_control::IgnitionROS2ControlPlugin::ISystemConfigure,
+  ignition_ros2_control::IgnitionROS2ControlPlugin::ISystemUpdate)
+
+IGNITION_ADD_PLUGIN_ALIAS(
+  ignition_ros2_control::IgnitionROS2ControlPlugin,
+  "ignition::gazebo::systems::IgnitionROS2ControlPlugin")
```

Comment:
I think we shouldn't be encouraging external projects to use the `ignition::gazebo` namespace. This should be reserved to functionality provided by the simulator.